pluton/spawn/containercache/sshcache.go

```diff
@@ -0,0 +1,278 @@
+package containercache
+
+import (
+	"bytes"
+	"fmt"
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/coreos/mantle/platform"
+	"github.com/coreos/mantle/util"
+)
+
+const (
+	cacheDir          = "/home/core/containercache"
+	downloadRetries   = 3
+	downloadRetryWait = 15 * time.Second
+	downloadTimeout   = 2 * time.Minute
+)
+
+var once sync.Once
+
+type ImageName struct {
+	Name   string
+	Engine string // docker or rkt
+}
+
+// StartBastionOnce will populate a bastion machine with the given containers
+// for rkt and docker. Since its best to run this only once per test suite, the
+// bastion machine should be in a separate platform.Cluster then the one used
+// in the test so it can be available for all tests in a given run. This
+// function is guarded with a sync.Once, so only the first within a test suite
+// it will have it run.
+func StartBastionOnce(bastion platform.Machine, names []ImageName) error {
+	var err error
+
+	f := func() {
+		err = startBastion(bastion, names)
+	}
+	once.Do(f)
+
+	return err
+}
+
+// Pull all containers cached by the bastion node into the docker and rkt store
+// of the given machines.
+func Pull(bastion platform.Machine, machines []platform.Machine) error {
+	if len(machines) < 0 {
+		return nil
+	}
+
+	// error if bastion is not setup
+	if err := StartBastionOnce(nil, nil); err != nil {
+		return fmt.Errorf("Must call StartBastionOnce before Pull")
+	}
+
+	if err := copyPublicKeys(bastion, machines); err != nil {
+		return fmt.Errorf("copying public keys: %v", err)
+	}
+
+	var wg sync.WaitGroup
+	var errors []error
+
+	wg.Add(len(machines))
+
+	for _, m := range machines {
+		go func(m platform.Machine) {
+			defer wg.Done()
+
+			if err := transferContainers(bastion, m); err != nil {
+				errors = append(errors, err)
+			}
+		}(m)
+	}
+	wg.Wait()
+
+	if len(errors) != 0 {
+		return fmt.Errorf("%s", errors)
+	}
+
+	return nil
+}
+
+func startBastion(bastion platform.Machine, names []ImageName) error {
+	if bastion == nil {
+		return fmt.Errorf("error starting containercache: bastion is nil")
+	}
+
+	// generate key-pair
+	out, err := bastion.SSH(`ssh-keygen -t ed25519 -N "" -f ./.ssh/bastion.key`)
```

Comment:
That makes sense. However, I think I prefer to keep things the way they are because this complicates use of the containercache package. Mainly, machines in the test cluster much be given a crafted userdata with the key. And since the logic is lifted out of the containercache pkg, it complicates use of its API since spawn must know to make sure bastion and test machines are setup already to SSH to each other.